sotodlib/io/load_book.py

```diff
@@ -0,0 +1,223 @@
+
+    """
+    # What dets do we have here and where do they belong?
+    _, src_idx, dest_idx = core.util.get_coindices(src.names, dets)
+    count = min(len(src.times) - src_offset, dest.shape[-1] - dest_offset)
+    if count < 0:
+        return len(src.times)
+    samp_slice = slice(src_offset, src_offset+count)
+    for i0, i1 in zip(src_idx, dest_idx):
+        dest[i1,dest_offset:dest_offset+count] = src.data[i0,samp_slice]
+    return src_offset + count
+
+def load_obs_book(db, obs_id, dets=None, prefix=None, samples=None,
+                  no_signal=None,
+                  **kwargs):
+    if any([v is not None for v in kwargs.values()]):
+        raise RuntimeError(
+            f"This loader function does not understand these kwargs: f{kwargs}")
+
+    if prefix is None:
+        prefix = db.prefix
+        if prefix is None:
+            prefix = './'
+
+    if no_signal is None:
+        no_signal = False  # from here, assume no_signal in [True, False]
+
+    # Regardless of what dets have been asked for (maybe none), get
+    # the list of detsets implicated in this observation.
+    c = db.conn.execute('select distinct DS.name, DS.det from detsets DS '
+                        'join files on DS.name=files.detset '
+                        'where obs_id=?', (obs_id,))
+    pairs = [tuple(r) for r in c.fetchall()]
+
+    # Now filter to only the dets requested.
+    if dets is None:
+        pairs_req = pairs
+    else:
+        pairs_req = [p for p in pairs if p[1] in dets]
+        # Use sets for this...
+        dets_req = [p[1] for p in pairs_req]
+        unmatched = [d for d in dets if d not in dets_req]
+        if len(unmatched):
+            raise RuntimeError("User requested invalid dets (e.g. %s) "
+                               "for obs_id=%s" % (unmatched[0], obs_id))
+    detsets_req = set(p[0] for p in pairs_req)
+
+    file_map = db.get_files(obs_id)
+    one_group = list(file_map.values())[0]  # [('file0', 0, 1000), ('file1', 1000, 2000), ...]
+    num_files = len(one_group)
+    book_root = os.path.split(one_group[0][0])[0]
+    
+    # Figure out how many samples we're loading.
+    sample_range = one_group[0][1], one_group[-1][2]
+    if samples is None:
+        samples = None, None
+    samples = list(samples)
+    if samples[0] is None:
+        samples[0] = sample_range[0]
+    if samples[1] is None:
+        samples[1] = sample_range[1]
+    elif samples[1] < 0:
+        samples[1] = sample_range[1] + samples[1]
+    samples[0] = min(max(0, samples[0]), sample_range[1])
+    samples[1] = min(max(samples), sample_range[1])
+
+    file_map = db.get_files(obs_id)
+
+    # Prepare the output structure.
+    aman = core.AxisManager(
+        core.LabelAxis('dets', [p[1] for p in pairs_req]),
+        core.OffsetAxis('samps', samples[1] - samples[0], samples[0], obs_id))
+    aman.wrap('primary', core.AxisManager(aman.samps))
+    aman.wrap('tes_bias', core.AxisManager(aman.samps))
```

Comment:
The docs do not generalize unambiguously to the multi-UFM case.  I'm happy to talk about a sensible generalization.

Reply:
Right, the generalization is a thing to work out in more detail....

What you have per detset isn't what we had per lvl2 axismanager for biases. Those are bias_line x samps arrays instead of single line arrays. At some points in the past I've been building `biases x dets` arrays for maps and things. So I have liked having that as a label axis and not a lot of single dim arrays. 

If we're going to want any arrays that are `biases x dets` in level 3 data then I we're going to want to extend those label axes to have names per detset or something. We also _definitely_ need to start building the caldbs for smurf things off of books / for use at level three because the standard formats we've had for bg_maps need updates. This is all probably work a discussion on one of the Thursday calls.